sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -6234,6 +6260,35 @@ pi_result piextGetDeviceFunctionPointer(pi_device Device, pi_program Program,
     ModIt++;
   }
 
+  // zeModuleGetFunctionPointer currently fails for all
```

Comment:
@jandres742 : is this known/expected?

Reply:
`zeModuleGetFunctionPointer` should work. We are not aware of any known bugs. 

@AidanBeltonS : would you mind sharing a test where you see this failing? I think it would be better to confirm whether the failure is legit before merging this patch with this comment.

Reply:
Yes I don't mind sharing a test. Unfortunately the regular usage of `piextGetDeviceFunctionPointer` before this PR appears to be somewhat deprecated. As the only test I can find being in DeprecatedFeatures/FunctionPointers/pass-fp-through-buffer.cpp.
This is causing me some issues calling this directly as a result the easiest way of demonstrating the behaviour is by outputting some behaviour from the level_zero PI.

This patch should be applied to my public branch `feature-piextGetDeviceFunctionPointer`.
Everytime `piextGetDeviceFunctionPointer` is called it prints the list of kernel names, the name of the kernel we want to find, and if the result is an invalid argument. It also prints the function pointer address to check if it is a nullptr.

<details>
  <summary>Patch</summary>
  
  ```
diff --git a/sycl/plugins/level_zero/pi_level_zero.cpp b/sycl/plugins/level_zero/pi_level_zero.cpp
index 2134b26651e7..7177337b581a 100644
--- a/sycl/plugins/level_zero/pi_level_zero.cpp
+++ b/sycl/plugins/level_zero/pi_level_zero.cpp
@@ -6248,6 +6248,29 @@ pi_result piextGetDeviceFunctionPointer(pi_device Device, pi_program Program,
     return PI_INVALID_PROGRAM_EXECUTABLE;
   }
 
+  {
+    // List all available kernels 
+    std::string PINames{""};
+    bool First = true;
+    _pi_program::ModuleIterator ModIt(Program);
+    while (!ModIt.Done()) {
+      uint32_t Count = 0;
+      ZE_CALL(zeModuleGetKernelNames, (*ModIt, &Count, nullptr));
+      std::unique_ptr<const char *[]> PNames(new const char *[Count]);
+      ZE_CALL(zeModuleGetKernelNames, (*ModIt, &Count, PNames.get()));
+      for (uint32_t I = 0; I < Count; ++I) {
+        PINames += (!First ? ";" : "");
+        PINames += PNames[I];
+        First = false;
+      }
+        ModIt++;
+    }
+    std::cout << "All Kernel Names:" << PINames << std::endl;
+    std::cout << "Kernel Name we want to find:" << FunctionName << std::endl;
+  }
+
+  std::cout << "reinterpret cast FunctionPointerRet: " << reinterpret_cast<void **>(FunctionPointerRet) << std::endl;
+
   // Search for the function name in each module.
   ze_result_t ZeResult = ZE_RESULT_ERROR_INVALID_FUNCTION_NAME;
   _pi_program::ModuleIterator ModIt(Program);
@@ -6255,6 +6278,8 @@ pi_result piextGetDeviceFunctionPointer(pi_device Device, pi_program Program,
     ZeResult = ZE_CALL_NOCHECK(
         zeModuleGetFunctionPointer,
         (*ModIt, FunctionName, reinterpret_cast<void **>(FunctionPointerRet)));
+    if (ZeResult == ZE_RESULT_ERROR_INVALID_ARGUMENT)
+      std::cout << "ZE_RESULT_ERROR_INVALID_ARUGMENT is returned\n";
     if (ZeResult != ZE_RESULT_ERROR_INVALID_FUNCTION_NAME)
       break;
     ModIt++;
  ```
  
</details>



Simplified test case from split-per-kernel.cpp simply creates one kernels then checks if it has one which is should posses 
and one that is shouldn't posses. `zeModuleGetFunctionPointer` is called for each kernel and the list of kernel names is also printed 
to show that the kernel can be seen.

<details>
  <summary>SYCLtest</summary>
  
  ```
 #include <CL/sycl.hpp>

class Kern1;
class Kern2;

int main() {
  cl::sycl::queue Q;
  int Data = 0;
  {
    cl::sycl::buffer<int, 1> Buf(&Data, cl::sycl::range<1>(1));
    cl::sycl::program Prg(Q.get_context());
    Prg.build_with_kernel_type<Kern1>();
    cl::sycl::kernel Krn = Prg.get_kernel<Kern1>();

    Prg.has_kernel<Kern1>();
    Prg.has_kernel<Kern2>();

    Q.submit([&](cl::sycl::handler &Cgh) {
      auto Acc = Buf.get_access<cl::sycl::access::mode::read_write>(Cgh);
      Cgh.single_task<Kern1>(Krn, [=]() { Acc[0] = 1; });
    });
  }
  assert(Data == 1);

  {
    cl::sycl::buffer<int, 1> Buf(&Data, cl::sycl::range<1>(1));
    cl::sycl::program Prg(Q.get_context());
    Prg.build_with_kernel_type<Kern2>();
    cl::sycl::kernel Krn = Prg.get_kernel<Kern2>();

    Q.submit([&](cl::sycl::handler &Cgh) {
      auto Acc = Buf.get_access<cl::sycl::access::mode::read_write>(Cgh);
      Cgh.single_task<Kern2>(Krn, [=]() { Acc[0] = 2; });
    });
  }
  assert(Data == 2);

  return 0;
}
  ```
  
</details>


Output
```
All Kernel Names:_ZTS5Kern1
Kernel Name we want to find:_ZTS5Kern1
reinterpret cast FunctionPointerRet: 0x7ffddfd48a48
ZE_RESULT_ERROR_INVALID_ARUGMENT is returned

All Kernel Names:_ZTS5Kern1
Kernel Name we want to find:_ZTS5Kern2
reinterpret cast FunctionPointerRet: 0x7ffddfd48a48
ZE_RESULT_ERROR_INVALID_ARUGMENT is returned

```

The first kernel does exist inside the module however `zeModuleGetFunctionPointer` returns an invalid argument error.
The second kernel also does not exist inside the module and the same error is returned.
From the docs https://spec.oneapi.io/versions/0.5.0/oneL0/api/function_ze__module_8h_1a46e5f2e11a0e9a19e8f974e05e436dff.html
this could be because any of the inputs are nullptrs. The most likely case for a bad input would be `FunctionPointerRet` after being cast to void** however from the output this is not returning a nullptr.

I am not very familiar with level_zero so let me know if I am passing some invalid arguments or am looking at out of date docs. Many thanks

Reply:
Thanks @AidanBeltonS .

@smaslov-intel : I would suggest to continue with this merge, and then file a bug so we can debug the issue related to L0.

Reply:
OK.

Reply:
@AidanBeltonS please open an issue to track this down.

Reply:
@AidanBeltonS : just a reminder to create issue for zeModuleGetFunctionPointer not returning the right value.

Reply:
Follow up issue: https://github.com/intel/compute-runtime/issues/460